### PR TITLE
async_unary_call: add a Destroy method, called by std::default_delete.

### PR DESCRIPTION
**Pull request by jacobsa**
This implements [gRFC L80](https://github.com/grpc/proposal/blob/a3a0e23b/L80-cpp-async-response-reader-destruction.md).

@soheilhy

### CODE REVIEW

**Comment by soheilhy** [comment]
thanks for the change @jacobsa!

@vjpai @ctiller folks, since this is considered an API change, should we bump the version up?

**Comment by jacobsa** [comment]
Oh yes sorry, that's part of the gRFC but I forgot to do it. Can you remind me which file I need to edit? Or is there more process than that?

**Comment by soheilhy** [comment]
I haven't done that since the build pipeline has been revamped. We used to bump the version `build.yaml`, but now the version is in the `build_handwritten.yaml`:

- https://github.com/grpc/grpc/blob/master/build_handwritten.yaml#L15-L19
- https://github.com/grpc/grpc/blob/master/BUILD#L87-L92

@jtattermusch could you please advise what's the best practice for bumping up the version? Thanks!

**Comment by jtattermusch** [comment]
> I haven't done that since the build pipeline has been revamped. We used to bump the version `build.yaml`, but now the version is in the `build_handwritten.yaml`:
> 
> * https://github.com/grpc/grpc/blob/master/build_handwritten.yaml#L15-L19
> * https://github.com/grpc/grpc/blob/master/BUILD#L87-L92
> 
> @jtattermusch could you please advise what's the best practice for bumping up the version? Thanks!

Currently, you need to bump the version both in build_handwritten.yaml and in BUILD (and they need to match). After that, you will run tools/buildgen/generate_projects.sh as normal.
(a bit more explanation: we've made BUILD the source of truth for vast majority of build metadata and now build.yaml is generated from the bazel BUILD. It's not perfect yet, so build_handwritten.yaml still has some "handwritten" metadata that we don't extract from BUILD yet - the plan is to remove build_handwritten.yaml in the future entirely).

**Comment by jacobsa** [comment]
Thanks. Could you please give me more specific instructions though? There are at least four different types of versions listed in those files, plus "guadalupe river park conservancy", whatever that means. (And it appears this silly phrase is supposed to be kept in sync with the "version", but it is out of date according to `g_stands_for.md`, as far as I can tell.)

Which of these am I supposed to update? Is there documentation on this?

**Comment by vjpai** [comment]
@jacobsa @jtattermusch This isn't a breaking change so it's only a minor version bump, and minor version bumps automatically at each 6-week release so I don't think that this concern is relevant, is it?

**Comment by jacobsa** [comment]
@vpai: The gRFC says it will be a version bump, because it is breaking for users that have subclassed the interface and depend on not actually being destroyed.

**Comment by jtattermusch** [comment]
I looked at the gRFC and it says "This change thus requires a major version change.".
I'm not sure what's the background of this proposal, but for bumping major version of gRPC C++, you'll need more approvals than just @soheilhy. 

Please collect approvals from @markdroth, @srini100, @a11r  and get an explicit approval that bumping major version of gRPC C++ is what we want to do (in my understanding we don't want that but as said, I don't have much context here).
Note that bumping major version of a language (gRPC C++ in this cas) is a lot of work, we only do it VERY rarely, it needs to satisfy strong requirements and the bar for getting it approved is pretty high.
See documentation on this topic: 
https://github.com/grpc/grpc/blob/master/doc/versioning.md#policy-for-updating-the-major-version-number

**Review by jtattermusch — CHANGES_REQUESTED** [review]
Based on the conversations, this seems pretty controversial due to potentially requiring a major version bump for gRPC C++.  I'm marking this as "request changes" until this gets sorted out and until we get explicit approvals from gRPC C++ owners.

**Comment by jacobsa** [comment]
Thanks. Can I ask you all to sort this out among yourselves ideally, and just let me know what to do? I was just trying to fix an API mistake; I really have no opinion about whether to change the version or not, and did a gRFC and included that only because I was asked to. (I also can't resist pointing out that you had 10 business days to tell me this as part of the gRFC process. It clearly said it was going to bump the version number, and my understanding is that was the time to object.)

**Comment by soheilhy** [comment]
The initial version of this pull request didn't need a major version bump but, as per @vjpai's comment on an earlier version of this PR, Aaron sent a gRFC which mentioned that we need to bump the major version.

The change itself is controversial.  The main question is whether we need a version bump. Adding @markdroth to see if he feels this needs a major version bump.  If not, we can probably modify the gRFC to remove the requirement for a major version bump.

**Comment by vjpai** [comment]
I didn't review the gRFC so I don't have a comment there. That said, I don't think this is a major version bump because we've said from the beginning that we don't support mocking of async calls (or at least only indicated that we support mocking of sync calls). So, anyone who was overriding this was using an unsupported API anyway and we don't have to maintain unsupported behavior. We've also said that we don't consider it a breaking API change if there was no public use of the API, and we got no evidence to the contrary during the gRFC process. I asked for a gRFC under the assumption that this was a standard minor-version additive API change because we do require gRFCs for those also.

**Comment by soheilhy** [comment]
Ah ok thanks @vjpai for the clarification! In that case, we won't need any major version bump. Let's wait for @markdroth's LGTM here and then submit the PR as is.

**Comment by veblush** [comment]
I agree that this doesn't need to bump the major version.

**Review by veblush — APPROVED** [review]

**Comment by jacobsa** [comment]
@vebush: Thanks! Are you (or anybody else) able to clear the "changes requested" bit? I don't think I have permission (or at least can't figure out how to do it).

**Comment by soheilhy** [comment]
I believe @jtattermusch needs to remove the change request, once he's happy with the outcome here.

**Review by soheilhy — APPROVED** [review]

**Review by jtattermusch — APPROVED** [review]
Approving to get rid of the "requested changes" status. The real approved should still come from @veblush  and @markdroth.
Please also update the gRFC proposal accordingly to explain why this needs / doesn't need a major version bump.

**Comment by jacobsa** [comment]
I have no opinion on whether  it needs/doesn't need a version bump. Can I ask @soheilhy or @veblush to do this, or at least to give me reasoning to use?

**Comment by soheilhy** [comment]
I think @vjpai put it best. Rephrasing his words:

This change is not API breaking change, and only extends an existing API. Since we don't officially support mocking of async calls, anyone who has been relying on overriding this class was in the "unsupported" territory. Thus, a minor version bump would suffice.

@jtattermusch Mark assigned the review to @veblush. Do we need to bother him for an approval?

**Comment by jacobsa** [comment]
@soheilhy: Okay thanks, see https://github.com/grpc/proposal/pull/242.

Could someone please give me specific instructions for exactly what to edit in order to change the minor version? It's not clear from the instructions above what a minor version is, and exactly which fields need to be edited.

**Comment by vjpai** [comment]
Don't edit anything for a minor version bump. The minor version automatically bumps at each new 6-week release, whether it actually needs to or not.

**Comment by jacobsa** [comment]
Thanks. In that case, I think there's nothing left here, right? Can someone please merge this?

**Comment by soheilhy** [comment]
I'll wait for @jtattermusch, @markdroth, or @veblush to merge this, since I don't know if Jan wants any further approval for this.

**Comment by jtattermusch** [comment]
Thanks for updating the proposal. I think with approvals from @veblush and @soheilhy  we should be fine, so I'm going to go ahead and merge.

**Comment by jacobsa** [comment]
Thanks!

**Pull request merged**
